WB-1063 - Implement FieldHeading Component

RivasCVA opened the pull request:
## Summary:
Added the `FieldHeading` private component under `wonder-blocks-form`.
This component is intended to provide a reusable handy label, description, and error element.

In the current situation, it is mainly intended to be used for `LabeledTextField`,
but later it may be easily used by other form fields.

Issue: https://khanacademy.atlassian.net/browse/WB-1063

## Test plan:
I added a temporary example to React Storybook (see generated link below).
You can view it under Form / FieldHeading.

#### NOTE: I will remove the `FieldHeading` example from React Storybook on the next PR.

---

##### Example 1: Showing Label and Description
<img width="419" alt="basic-1" src="https://user-images.githubusercontent.com/60367213/121592318-22f19300-ca00-11eb-811a-445ead29704d.png">

##### Example 2: Showing Error
<img width="420" alt="basic-error-1" src="https://user-images.githubusercontent.com/60367213/121592365-34d33600-ca00-11eb-85b7-ca12208147f7.png">

##### Example 3: Passing Custom Typography Labels
<img width="420" alt="basic-custom" src="https://user-images.githubusercontent.com/60367213/121592403-40bef800-ca00-11eb-885b-78c18d9938e5.png">

RivasCVA commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+                )}
+                {description && <Strut size={Spacing.xxxSmall_4} />}
+
+                <Strut size={Spacing.xSmall_8} />
+                {field}
+
+                {error && <Strut size={Spacing.small_12} />}
+                {error && this.isString(error) ? (
+                    <LabelSmall style={[styles.error]}>{error}</LabelSmall>
+                ) : (
+                    error
+                )}
```
Would there be a better way to ensure there is space reserved below the "field" for when there is potential for an error message. My thinking is that we would not want things below this component to be "pushed down" when the error message appears.

RivasCVA commented:
Just to make sure: I am not to export this `FieldHeading` component under `index.js` since it is intended to be a private component, correct?

jandrade commented:
> Just to make sure: I am not to export this `FieldHeading` component under `index.js` since it is intended to be a private component, correct?

That's right :)

RivasCVA commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+    render(): React.Node {
+        const {field, label, description, error} = this.props;
+        return (
+            <View>
+                {label && this.isString(label) ? (
+                    <LabelMedium style={[styles.label]}>{label}</LabelMedium>
+                ) : (
+                    label
+                )}
+                {<Strut size={Spacing.xxxSmall_4} />}
+
+                {description && this.isString(description) ? (
+                    <LabelSmall style={[styles.description]}>
+                        {description}
+                    </LabelSmall>
+                ) : (
+                    description
+                )}
+                {description && <Strut size={Spacing.xxxSmall_4} />}
+
+                <Strut size={Spacing.xSmall_8} />
+                {field}
+
+                {error && <Strut size={Spacing.small_12} />}
+                {error && this.isString(error) ? (
+                    <LabelSmall style={[styles.error]}>{error}</LabelSmall>
+                ) : (
+                    error
+                )}
```
Also, is it best to use `<Strut />` here to correctly space the labels from each other and the field, or using `padding` in the corresponding elements would be better?

jandrade replied:
I mentioned this in one of my comments. `Strut` is perfect for these use case :)

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+                {<Strut size={Spacing.xxxSmall_4} />}
+
+                {description && this.isString(description) ? (
+                    <LabelSmall style={[styles.description]}>
```
It would be good to add the `tag="label"` prop here to add a semantic meaning to this element.

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+        return typeof item === "string";
+    };
+
+    render(): React.Node {
```
As we are going to generate IDs to connect the label with the field, it would be good to use `IDProvider` here to handle that logic for us.

You can see an example here: https://deploy-preview-1140--wonder-blocks.netlify.app/#!/IDProvider/1

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+                {description && <Strut size={Spacing.xxxSmall_4} />}
+
+                <Strut size={Spacing.xSmall_8} />
+                {field}
```
Are you planning to add `aria-invalid` and `aria-describedby` when you implement `LabeledTextField`?

RivasCVA replied:
Good point! I will be testing some of those attributes on the changes I will be making shortly.

jandrade reviewed (changes requested):
Thanks for adding this component, it will help with future implementations!.

I'm requesting some changes that will be necessary to provide better tools for testing the implementation on parent components, adding IDProvider to connect elements and adding some ARIA roles to cover accessibility.

RivasCVA commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,90 @@
+
+                {error && <Strut size={Spacing.small_12} />}
+                {error && this.isString(error) ? (
+                    <LabelSmall style={[styles.error]}>{error}</LabelSmall>
```
> According to the design specs, this should be using `LabelMedium` by default.

It I'm looking that this Figma file ([Figma1](https://www.figma.com/file/iIAS9PG3vA0lbcJbtTPQIg/Form-Field-Component-Architecture?node-id=10%3A40)) and this other one ([Figma2](https://www.figma.com/file/VbVu3h2BpBhH80niq101MHHE/Wonder-Blocks-(Web)?node-id=495%3A0)), and both of them have `LabelSmall` for the default error. Are you perhaps looking at another Figma file I might not have seen?

RivasCVA replied:
> Another thing to have in mind is that this component should include `role="alert"` to notify that the field contains an error. (Apologies for missing this in the implementation spec).
> 
> More info here: https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html

TIL about `role="alert"`. Thanks for the link! I'll be sure to add that.

jandrade replied:
> > According to the design specs, this should be using `LabelMedium` by default.
> 
> It I'm looking that this Figma file ([Figma1](https://www.figma.com/file/iIAS9PG3vA0lbcJbtTPQIg/Form-Field-Component-Architecture?node-id=10%3A40)) and this other one ([Figma2](https://www.figma.com/file/VbVu3h2BpBhH80niq101MHHE/Wonder-Blocks-(Web)?node-id=495%3A0)), and both of them have `LabelSmall` for the default error. Are you perhaps looking at another Figma file I might not have seen?

Ohh you're right! I think I saw that in another spec, but these are the official ones, so let's keep `LabelSmall`.

RivasCVA commented:
New Changes:
* Took out the rendering of the different elements into separate functions to make the code more readable.
* Added the `testId` prop.
  * Passed it to the `label`, `description`, and `error` elements formatted as `testid-label`, `testid-description`, and `testid-error` correspondingly.
* Added the `id` prop.
  * Used it to be able to match the label's `htmlFor` attribute with the input's `id` attribute (formatted as `id-field`).
  * Used it to be able to match the error's `id` attribute to the input's `aria-describedby` attribute (formatted as `id-error`).
* Added `tag="label"` to label.
* Added `role="alert"` to error.
* Fixed suggested changes to the unit test regarding placing the code under "Act" section.
* Added more unit tests to cover the new added props (current coverage is 100%).

jandrade commented on packages/wonder-blocks-form/src/components/__tests__/field-heading.test.js:
```diff
@@ -0,0 +1,157 @@
+        );
+
+        // Assert
+        const label = wrapper.find(`[data-test-id="${testId}-label"]`);
```
Nice, thanks for adding these `testId` tests :)

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,148 @@
+
+        return (
+            <View>
+                {this.renderLabel()}
+                {this.maybeRenderDescription()}
+                <Strut size={Spacing.xSmall_8} />
+                {field}
+                {this.maybeRenderError()}
```
I really like how this is more readable!

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,148 @@
+            <React.Fragment>
+                <Strut size={Spacing.small_12} />
+                {typeof error === "string" ? (
+                    <LabelSmall
```
Just one minor change to match the design specs:
```suggestion
                    <LabelMedium
```

jandrade commented on packages/wonder-blocks-form/src/components/field-heading.stories.js:
```diff
@@ -0,0 +1,114 @@
+            field={
+                <TextField
+                    id={`${uniqueId}-field`}
+                    aria-describedby={`${uniqueId}-error`}
+                    aria-invalid={errorMessage ? "true" : "false"}
```
Great, thanks for adding the examples here! Looking forward to see this implementation in `LabeledTextField`.

jandrade reviewed (approved):
Looks great! 💪 thanks for refactoring the component and making all the changes, I really appreciate it :)

somewhatabstract commented on packages/wonder-blocks-form/src/components/field-heading.js:
```diff
@@ -0,0 +1,148 @@
+    maybeRenderError(): ?React.Node {
+        const {error, id, testId} = this.props;
+
+        if (!error) {
+            return null;
+        }
```
Yes! Love this style of optional rendering.

somewhatabstract reviewed (approved):
LGTM
